source/module_rdmft/rdmft_tools.cpp

```diff
@@ -0,0 +1,414 @@
+//==========================================================
+// Author: Jingang Han
+// DATE : 2024-03-11
+//==========================================================
+
+#include "module_rdmft/rdmft_tools.h"
+#include "module_hamilt_pw/hamilt_pwdft/global.h"
+// used by class Veff_rdmft
+#include "module_base/tool_title.h"
+#include "module_base/timer.h"
+#include "module_hamilt_general/module_xc/xc_functional.h"
+#include "module_elecstate/potentials/H_Hartree_pw.h"
+#include "module_elecstate/potentials/pot_local.h"
+#include "module_elecstate/potentials/pot_xc.h"
+#include "module_hamilt_pw/hamilt_pwdft/structure_factor.h"
+
+#include <iostream>
+#include <cmath>
+#include <complex>
+#include <fstream>
+#include <sstream>
+
+
+namespace rdmft
+{
+
+
+template <>
+void conj_psi<double>(psi::Psi<double>& wfc) {}
+
+
+template <>
+void HkPsi<double>(const Parallel_Orbitals* ParaV, const double& HK, const double& wfc, double& H_wfc)
+{
+    const int one_int = 1;
+    const double one_double = 1.0;
+    const double zero_double = 0.0;
+    const char N_char = 'N';
+    const char C_char = 'C';
+
+#ifdef __MPI
+    const int nbasis = ParaV->desc[2];
+    const int nbands = ParaV->desc_wfc[3];
+
+    //because wfc(bands, basis'), H(basis, basis'), we do wfc*H^T(in the perspective of cpp, not in fortran). And get H_wfc(bands, basis) is correct.
+    pdgemm_( &C_char, &N_char, &nbasis, &nbands, &nbasis, &one_double, &HK, &one_int, &one_int, ParaV->desc,
+        &wfc, &one_int, &one_int, ParaV->desc_wfc, &zero_double, &H_wfc, &one_int, &one_int, ParaV->desc_wfc );
+#endif
+
+}
+
+
+template <>
+void psiDotPsi<double>(const Parallel_Orbitals* ParaV, const Parallel_2D& para_Eij_in,
```

Comment:
I will not comment on your code structure, but care much about readability. As far as I understood, this function actually does the inner product of two Hilbert vector in representation of some basis, the one is psi and the other might be the psi after act by some operator? So it is possible to pass H|psi> to wfc, and |psi> to H_wfc? also the "H" is not directly/specifically the Hamiltonian operator, but actually any operator?

Reply:
seems you are doing two things in one function, the first is to calculate the inner product of bra and op_ket, the other is to extract diagonal elements in bra_op_ket. So the function you named PsiDotPsi is a quite special function that should be properly treated. My suggestion is, you can separate your function into two smaller functions, the first named cal_bra_op_ket, and the second named _diagonal_in_serial. Then you get the wfcHwfc by _diagonal_in_serial(cal_bra_op_ket(..., bra, ..., op_ket, ...)).
Do you think this is better?
You can then add unittests on them to ensure the correctness and stability.

Reply:
Yes, your solution is better, I will modify it.
But considering that nbands will be relatively large in the RDMFT method, I still plan to use the formal parameter Dmn(k) of the reference variable instead of the return value in the function cal_bra_op_ket(…), is this okay?
Also, I am a little curious about why you used "_" as the first letter when naming _diagonal_in_serial()

Reply:
@JGHan7 I think it is acceptable.
the reason why I add one "_" ahead is to indicate this is a private or context-specific implementation (not designed to be generally used elsewhere). Now there are new implementations combining this nomenclature with anonymous namespace (a relatively advanced C++ feature), to make the interface of code very clean. For an instance, please see: 
https://github.com/deepmodeling/abacus-develop/blob/6df5c6dc254c204f8d6acfa7f01cf5f4de2ca1a0/source/module_base/grid/batch.cpp#L41

Reply:
OK, thank you for your detailed explanation.